### ISSUE TITLE
Toggle unit state - `start` if stopped and `stop` if running

### ISSUE THREAD

**Issue by tuxiqae** [systemd/systemd#19256]
**Is your feature request related to a problem? Please describe.**
The request is not related to a problem.

**Describe the solution you'd like**
 I think that it could be really efficient to have a single command which will `start` a unit if it is stopped and `stop` a unit if it is running.
Such a solution should be used for daily use of the terminal and not for script-writing.

**Describe alternatives you've considered**
Alternatives may include writing custom scripts which check the state of the unit and acts accordingly.

**The systemd version you checked that didn't have the feature you are asking for**
248.
I have also checked the man and the repository and couldn't find any code or issues regarding such feature.

**Comment by amarjargal** [comment]
Hello everyone. I'm an outreachy applicant hoping to make some contributions to systemd. I'd like to work on this issue if it's still available.

**Comment by tuxiqae** [comment]
Still is as far as I'm aware

**Comment by amarjargal** [comment]
I tried to tackle this issue in PR #23077. Please let me know if this solution is OK.

**Comment by poettering** [comment]
Hmm, anyone knows what the precise usecase for this is? i.e. why would you want to an issue where you can't guess the possible outcome? Would be good to shed some light on this.

**Comment by behrmann** [comment]
I would like to have something like this mostly during development, where I start a service, see that something isn't working the way I want it to, stop the service, tweak the config and/or recompile, start the service again. It's not like this is something I couldn't live without, I have `systemctl start foo.service` and `systemctl stop foo.service` in my shell's history and I won't do much typing anyway, but it would still be a quality of life thing.

I do think, though, it would make sense if this command were verbose by default, e.g.
```
$ systemctl toggle foo.service
starting foo.service
$ systemctl toggle foo.service
stoppping foo.service
```

**Comment by james-lawrence** [comment]
@behrmann that sounds like 'restart' does what you want.

**Comment by tuxiqae** [comment]
> 

But it doesn't, `restart` `stop`s and then **immediately** `start`s the service.
Toggle does one or the other at a time, not both.

**Comment by tuxiqae** [comment]
A better example might be the following:

```
$ systemctl toggle foo.service
starting foo.service

# ... Do something ...

$ systemctl toggle foo.service
stoppping foo.service
```

**Comment by james-lawrence** [comment]
not really. doesn't explain why going from unknown state -> unknown state is useful or desirable.

- if you want to ensure a service is running but don't want to do anything if its already running -> use systemctl start.
- if you want to ensure a service is running and reset it -> use systemctl restart.
- if you want to ensure a service is not running -> use systemctl stop.

I'm not seeing a situation where it'd be desirable for one to simply invert the state of a service without the state of the service being known in advance prior to performing actions.

**Comment by james-lawrence** [comment]
@bjesus you're still assuming a known state of the service. which means you can simply use the correct command without systemd implementing a toggle command.

**Comment by bjesus** [comment]
the fact that a UI displays the current state doesn't mean that I have that kind of access to it. I can link the clicking functionality to a script that re-checks the state and then executes the right command, but the whole point of the the `toggle` option is to remove the need for that. it's a shortcut... just like `restart` is essentially `stop & start`.

I thought the speakers on/off key is a good example. sure, I can _hear_ that the speakers are on because the music is playing, or I know their off because my video doesn't have sound. but it doesn't mean that the button I set up in the UI knows what command to execute, and so ideally i could just tell it to toggle pipewire.

**Comment by james-lawrence** [comment]
> the fact that a UI displays the current state doesn't mean that I have that kind of access to it.

you then go on to explain that you do actually have access to the checking the state. meaning you can easily determine on click which command to use.

**Comment by nsvedberg** [comment]
This would be desirable for me, it would make some WM configuration easier.

**Comment by nullref** [comment]
This is a naive implementation of service toggling:

```shell
if systemctl --user is-active --quiet toggle.service
then systemctl --user stop toggle.service
else systemctl --user start toggle.service
fi
```

Implementing this logic in _systemd_ would save one execution of `systemctl` while possibly also getting rid of the race condition in which the `stop` or `start` command is called unnecessarily.  To save the natural environment we must save energy at all costs.

**Comment by hfytr** [comment]
I have a usecase: i use kmonad as a service for things like home row mods. If someone else uses my computer, I would like to turn this off as they probably won't be used to it. Thus something like `sudo systemctl toggle kmonad.service` would be useful to map to a keyboard shortcut.

**Comment by mabequinho** [comment]
> > Hmm, anyone knows what the precise usecase for this is? i.e. why would you want to an issue where you can't guess the possible outcome? Would be good to shed some light on this.
> 
> I actually checked if such thing exists for my use case before I've seen this issue.
> 
> This can be useful for user-level units, for example I use SwayWM and have an ambient rain sounds toggle bound to `if systemctl --user is-active foo; then systemctl --user stop foo; else systemctl --user start foo; fi` so I can toggle it (it just has `play` command with an mp3 file). Would be nicer to just bind `systemctl --user toggle foo`.
> 
> Same thing I have for recording with `wf-recorder` to record screen.

I had the same issue. It is possible to toggle background noise using wireplumber. `wpctl set-mute ID toggle`, you can get the ID running `wpctl status`. 

That's ridiculously off-topic, sorry for the reply.

**Comment by jdh8** [comment]
I have a use case.  I want to map `systemctl --user toggle some.service` to a keyboard shortcut.  Given the scarcity of keys, I don't want two separate keys for starting and stopping.

**Comment by GreyXor** [comment]
> I have a use case. I want to map `systemctl --user toggle some.service` to a keyboard shortcut. Given the scarcity of keys, I don't want two separate keys for starting and stopping.

`systemctl is-active --quiet your-service && systemctl stop your-service || systemctl start your-service`

**Comment by mooreye** [comment]
> > I have a use case. I want to map `systemctl --user toggle some.service` to a keyboard shortcut. Given the scarcity of keys, I don't want two separate keys for starting and stopping.
> 
> `systemctl is-active --quiet your-service && systemctl stop your-service || systemctl start your-service`
> 
> This one-liner toggles a systemd service: if your-service is running, it stops it; if it’s not running, it starts it.
> 
> It uses systemctl `is-active` to check the current status, and shell short-circuit logic (&& / ||) to decide what to do next. Great for simple toggle scripts or keybindings. Feel free to add `--user` if needed

This is not equivalent to a proper `if ... then ... else` statement. If unit is active and stopping it fails, it will start it again.

https://mywiki.wooledge.org/BashPitfalls#cmd1_.26.26_cmd2_.7C.7C_cmd3

**Comment by sinni800** [comment]
I support this issue (which would really not take much code) for the reason of keyboard shortcuts. For user services that are always visible, but you don't want to use more buttons, you can reduce inputs by just inferring state from what you see and knowing what a button press at this time will do.

Think about that instead of me wanting to put my finger somewhere else, I want to change behavior by doing it at a different, observable time. I'm varying my input in the time-domain rather than the physical one, and why not support that?
